test/ec_comp_test.c

```diff
@@ -0,0 +1,2310 @@
+            return "?";
+    }
+}
+
+static const char *decode_point_form(int f) {
+    switch (f) {
+        case POINT_CONVERSION_UNCOMPRESSED:
+            return "uncompressed";
+        case POINT_CONVERSION_COMPRESSED:
+            return "compressed";
+        case POINT_CONVERSION_HYBRID:
+            return "hybrid";
+        default:
+            return "?";
+    }
+}
+#endif
+
+static int init_curves(void)
+{
+    size_t i;
+
+    num_known_curves = EC_get_builtin_curves(NULL, 0);
+    known_curves = OPENSSL_malloc(num_known_curves*sizeof(*known_curves));
+    if (   known_curves == NULL
+        || EC_get_builtin_curves(known_curves, num_known_curves) == 0)
+        goto fail;
+
+    known_curve_names = OPENSSL_malloc(sizeof(char *)*num_known_curves);
+    if (known_curve_names == NULL)
+        goto fail;
+
+    for (i=0; i<num_known_curves; ++i) {
+        known_curve_names[i] = OBJ_nid2sn(known_curves[i].nid);
+        if (!strcmp(known_curve_names[i], "Oakley-EC2N-3"))
+            nid_Oakley_EC2N_3 = known_curves[i].nid;
+        if (!strcmp(known_curve_names[i], "Oakley-EC2N-4"))
+            nid_Oakley_EC2N_4 = known_curves[i].nid;
+    }
+
+    return 1;
+
+fail:
+    OPENSSL_free(known_curves);
+    known_curves = NULL;
+    OPENSSL_free(known_curve_names);
+    known_curve_names = NULL;
+    return 0;
+}
+
+static EC_KEY *generate_new_key(size_t curve_i, int asn1_flag, int point_form, struct default_config_st *dcfg)
+{
+    EC_KEY *k;
+    EC_GROUP *g;
+
+    g = EC_GROUP_new_by_curve_name(known_curves[curve_i].nid);
+    if (g == NULL)
+        return NULL;
+
+#if 0
+    if (asn1_flag >= 0)
+        EC_GROUP_set_asn1_flag(g, asn1_flag);
+    if (point_form >= 0)
+        EC_GROUP_set_point_conversion_form(g, point_form);
+#endif
+
+    k = EC_KEY_new();
+    if (k == NULL) {
+        EC_GROUP_free(g);
+        return NULL;
+    }
+
+    if (EC_KEY_set_group(k, g) == 0) {
+        EC_KEY_free(k);
+        EC_GROUP_free(g);
+        return NULL;
+    }
+
+    EC_GROUP_free(g);
+
+    if (EC_KEY_generate_key(k) == 0) {
+        EC_KEY_free(k);
+        return NULL;
+    }
+
+    dcfg->asn1_flag  = EC_GROUP_get_asn1_flag(EC_KEY_get0_group(k));
+    dcfg->point_form = EC_GROUP_get_point_conversion_form(EC_KEY_get0_group(k));
+
+    if (asn1_flag >= 0)
+        EC_KEY_set_asn1_flag(k, asn1_flag);
+    if (point_form >= 0)
+        EC_KEY_set_conv_form(k, point_form);
+
+    return k;
+}
+
+static EC_KEY *get_existing_key(size_t curve_i, size_t param_format_i, size_t comp_format_i, struct default_config_st *dcfg)
+{
+    EC_KEY *k = NULL;
+    FILE *f = NULL;
+    BIO *b = NULL;
+    char filename[256];
+    snprintf(filename, sizeof(filename), "test/recipes/15-test_ec_comp_data/%s-%s-%s.priv",
+        known_curve_names[curve_i],
+        param_formats_i[param_format_i] < 0 ? "named_curve" : param_formats[param_format_i],
+        comp_formats_i[comp_format_i] < 0 ? "uncompressed" : comp_formats[comp_format_i]);
```

Comment:
The "unspecified" cases should be handled with asserts in `get_existing_key`: they do not make sense, so we just make sure callers don't accidentally pass nonsense.